src/include/daos/pool.h

```diff
@@ -12,13 +12,15 @@
 #ifndef __DD_POOL_H__
 #define __DD_POOL_H__
 
+#include <daos_types.h>
+#include <daos_prop.h>
+#include <daos_pool.h>
+#include <daos_task.h>
+
 #include <daos/common.h>
 #include <gurt/hash.h>
 #include <daos/pool_map.h>
 #include <daos/rsvc.h>
-#include <daos/tse.h>
-#include <daos_types.h>
-#include <daos_pool.h>
```

Comment:
This file should still `include <daos/tse.h>` right?

Reply:
On what basis?  That in uses some of the defines in daos/tse.h I assume?  We have to assume that every file that includes this already includes daos/tse.h via some route.

It's easy enough to run my new unit test on the private headers as well, I'll update this PR with a few more code changes to allow all of them to be included but I won't enable this by default as it means the test would need to access abt.h and also we have daos/qat.h which we don't seem to include from anywhere today.

I've also found that using the macros from gurt/debug.h depends on gurt/common.h but you can't include gurt/common.h before gurt/debug.h so you have to include gurt/debug.h then gurt/common.h or just gurt/common.h

Reply:
> On what basis? That in uses some of the defines in daos/tse.h I assume?

Right.

>  We have to assume that every file that includes this already includes daos/tse.h via some route.

That works accidentally, but not in the general sense. `daos/pool.h` now has a silent dependency on `daos/tse.h`.


> the macros from gurt/debug.h depends on gurt/common.h

That issue is exactly the kind of issue I'm trying to avoid. If "the macros from gurt/debug.h depends on gurt/common.h" then `gurt/debug.h` should include `gurt/common.h`

Reply:
> > On what basis? That in uses some of the defines in daos/tse.h I assume?
> 
> Right.
> 
> > We have to assume that every file that includes this already includes daos/tse.h via some route.
> 
> That works accidentally, but not in the general sense. `daos/pool.h` now has a silent dependency on `daos/tse.h`.

Having testing more now I don't think it does, `daos/pool.h` obviously pulls in `daos/tse.h` from one of the header files that it includes.  We could no doubt do some pruning of includes from c files but getting the headers right is the first step towards that.

> > the macros from gurt/debug.h depends on gurt/common.h
> 
> That issue is exactly the kind of issue I'm trying to avoid. If "the macros from gurt/debug.h depends on gurt/common.h" then `gurt/debug.h` should include `gurt/common.h`

It can't, `gurt/common.h` includes `gurt/debug.h`.  I'll deal with that another day.

Reply:
This is what I mean. Currently, `pool.h` has an _**implicit**_ dependency on `tse.h`. If, for whatever reason, `something.h` gets refactored and doesn't need `tse.h` anymore, then `pool.h` will be missing an include. And since `pool.h` already includes `tse.h`, I don't see the value in removing it.
![include](https://github.com/daos-stack/daos/assets/13281535/937cb58b-5b5f-42be-a4e0-7265b788f1e1)

Reply:
> We could no doubt do some pruning of includes from c files

I don't think the goal should be to reduce the includes to the minimal possible that allows us to build each header independently. We should include what we need in the headers that need it. Otherwise, we end up breaking unrelated files if we reorganize, like in my example above.